File: 3341.find-minimum-time-to-reach-last-room-i.py
#
# @lc app=leetcode id=3341 lang=python3
#
# [3341] Find Minimum Time to Reach Last Room I
#

# @lc code=start
from typing import List
import heapq

class Solution:
    def minTimeToReach(self, moveTime: List[List[int]]) -> int:
        n = len(moveTime)
        m = len(moveTime[0])
        # Dijkstra's algorithm: (time, i, j)
        pq = []
        heapq.heappush(pq, (0, 0, 0))
        dist = [[float('inf')] * m for _ in range(n)]
        dist[0][0] = 0
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        while pq:
            t, i, j = heapq.heappop(pq)
            if i == n-1 and j == m-1:
                return t
            if t > dist[i][j]:
                continue
            for di, dj in dirs:
                ni, nj = i + di, j + dj
                if 0 <= ni < n and 0 <= nj < m:
                    # The earliest time we can enter (ni,nj) is max(t+1, moveTime[ni][nj])
                    nt = max(t + 1, moveTime[ni][nj])
                    if nt < dist[ni][nj]:
                        dist[ni][nj] = nt
                        heapq.heappush(pq, (nt, ni, nj))
        return dist[n-1][m-1]
# @lc code=end

File: 3341.find-minimum-time-to-reach-last-room-i.cpp
//
// @lc app=leetcode id=3341 lang=cpp
//
// [3341] Find Minimum Time to Reach Last Room I
//

// @lc code=start
class Solution {
public:
    int minTimeToReach(vector<vector<int>>& moveTime) {
        int n = moveTime.size();
        int m = moveTime[0].size();
        
        vector<vector<int>> dist(n, vector<int>(m, INT_MAX));
        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<tuple<int, int, int>>> pq;
        
        dist[0][0] = 0;
        pq.push({0, 0, 0});
        
        int dx[] = {0, 0, 1, -1};
        int dy[] = {1, -1, 0, 0};
        
        while (!pq.empty()) {
            auto [time, x, y] = pq.top();
            pq.pop();
            
            if (x == n - 1 && y == m - 1) {
                return time;
            }
            
            if (time > dist[x][y]) {
                continue;
            }
            
            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                
                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    int arrivalTime = max(time, moveTime[nx][ny]) + 1;
                    
                    if (arrivalTime < dist[nx][ny]) {
                        dist[nx][ny] = arrivalTime;
                        pq.push({arrivalTime, nx, ny});
                    }
                }
            }
        }
        
        return dist[n-1][m-1];
    }
};
// @lc code=end